### PR TITLE
wxGUI wx.metadata: Fix launch g.gui.metadata, g.gui.cswbrowser

### PR DESCRIPTION
**Pull request by tmszi**
1. wxGUI `g.gui.metadata`:

g_gui_metadata

with csw dialog:

g_gui_metadata_with_csw_dialog

Exporting metadata reports to format PDF:

I changed the api to get static map img from google maps api to [osm api](https://github.com/jperelli/osm-static-maps) (sometimes the service `https://osm-static-maps.herokuapp.com` is unavailable for a short time). Currently only static map img url query `zoom` param argument is not correctly derived.

Example raster Basins maps (nc_basic_spm_grass7 location) url to get [static map img ](https://osm-static-maps.herokuapp.com/?geojson=[{'type':'Feature','geometry':{'type':'Polygon','coordinates':[[[-78.77462049,35.68792712],[-78.60889171,35.6875073],[-78.60830318,35.80918894],[-78.77428134,35.80960938],[-78.77462049,35.68792712]]]}}]&center=-78.69152414,35.74858701&zoom=12&size=200x200`) (Geographic Location section in the pdf doc):
```https://osm-static-maps.herokuapp.com/?geojson=[{'type':'Feature','geometry':{'type':'Polygon','coordinates':[[[-78.77462049,35.68792712],[-78.60889171,35.6875073],[-78.60830318,35.80918894],[-78.77428134,35.80960938],[-78.77462049,35.68792712]]]}}]&center=-78.69152414,35.74858701&zoom=12&size=200x200```

2. wxGUI `g.gui.cswbrowser`:

g_gui_cswbrowser

Additional info:

Tested with:

```
wx.version()
4.0.7.post2 gtk3 (phoenix) wxWidgets 3.0.4'
```

### CODE REVIEW

**Comment by ninsbl** [comment]
Cool. I guess this fixes: https://trac.osgeo.org/grass/ticket/3558 ?!

**Comment by tmszi** [comment]
> Cool. I guess this fixes: https://trac.osgeo.org/grass/ticket/3558 ?!

Thanks. First error message from this ticket related with old py2 wxPython version 3.0.0. I tested addon with py3 wxPython `4.0.7.post2 gtk3 (phoenix) wxWidgets 3.0.4'` version. Second message is `oswlib` py module warning message and still appear after start `g.gui.metadata` module.

```
/usr/lib64/python3.6/site-packages/owslib/iso.py:117: FutureWarning: the .identification and .serviceidentification properties will merge into .identification being a list of properties.  This is currently implemented in .identificationinfo.  Please see https://github.com/geopython/OWSLib/issues/38 for more information
  FutureWarning)
/usr/lib64/python3.6/site-packages/owslib/iso.py:482: FutureWarning: The .keywords and .keywords2 properties will merge into the .keywords property in the future, with .keywords becoming a list of MD_Keywords instances. This is currently implemented in .keywords2. Please see https://github.com/geopython/OWSLib/issues/301 for more information
  FutureWarning)
```

**Comment by neteler** [comment]
Wow, you did a huge work, @tmszi !

I'll try it out following https://grasswiki.osgeo.org/wiki/ISO/INSPIRE_Metadata_Support

**Comment by tmszi** [comment]
There is another error during installation:

` g.extension operation=add extension=wx.metadata url=/home/tomas/grass-addons/grass7/gui/wxpython/wx.metadata`

```
</home/tomas/grass-addons/grass7/gui/wxpython/wx.metadata> (be patient)...
Compiling...
Traceback (most recent call last):
  File "/tmp/grass7-tomas-18402/tmpkyio8kpk/wx.metadata/scripts/db.csw.admin", line 248, in <module>
    from mdlib.cswutil import *
ModuleNotFoundError: No module named 'mdlib'
make[1]: *** [/usr/lib64/grass79/include/Make/Html.make:14: db.csw.admin.tmp.html] Error 1
Traceback (most recent call last):
  File "/tmp/grass7-tomas-18402/tmpkyio8kpk/wx.metadata/scripts/g.gui.cswbrowser", line 22, in <module>
    from mdlib.cswlib import CSWBrowserPanel, CSWConnectionPanel
ModuleNotFoundError: No module named 'mdlib'
make[1]: *** [/usr/lib64/grass79/include/Make/Html.make:14: g.gui.cswbrowser.tmp.html] Error 1
Traceback (most recent call last):
  File "/tmp/grass7-tomas-18402/tmpkyio8kpk/wx.metadata/scripts/g.gui.metadata", line 44, in <module>
    from mdlib import mdgrass
ModuleNotFoundError: No module named 'mdlib'
```

**Comment by tmszi** [comment]
I fixed all the errors during the installation of the add-on. It is currently possible to install the add-on correctly.

**Comment by tmszi** [comment]
I fixed all the errors during the installation of the add-on. It is currently possible to install the add-on correctly.

File: grass7/gui/wxpython/wx.metadata/g.gui.metadata/g.gui.metadata.py [hunk 1]
```diff
@@ -21,41 +21,40 @@
...
+from mdlib.mdpdffactory import PdfCreator
 
-from lxml import etree
 import wx
-from wx.lib.buttons import ThemedGenBitmapTextButton as BitmapBtnTxt
 from wx import SplitterWindow
+from wx.lib.buttons import ThemedGenBitmapTextButton as BitmapBtnTxt
 
-#from pydispatch import dispatcher
+# from pydispatch import dispatcher
```
Comment (neteler) [inline grass7/gui/wxpython/wx.metadata/g.gui.metadata/g.gui.metadata.py:56]:
```suggestion
# from pydispatch import dispatcher

#%module
#% description: Graphical ISO/INSPIRE metadata editor.
#% keyword: general
#% keyword: GUI
#% keyword: metadata
#%end
```

Silence: "WARNING: Bug in UI description. Missing module description"

--> might be extended

**Comment by neteler** [comment]
Thanks for your enormous efforts to update this suite!

Just two minor warnings I got, which might be fixed as well:

```
python3 -t -m py_compile /home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdeditorfactory.py
/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdeditorfactory.py:608: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if value is None or value is '':
python3 -t -m py_compile /home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdutil.py
/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdutil.py:159: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if md.identification.identtype is '':
```

File: grass7/gui/wxpython/wx.metadata/g.gui.cswbrowser/g.gui.cswbrowser.py [hunk 2]
```diff
@@ -9,17 +9,18 @@
...
 import grass.script as grass
+from grass.script.setup import set_gui_path
+
+grass.utils.set_path(modulename='wx.metadata', dirname='mdlib', path='..')
 
-from grass.pygrass.utils import set_path
-set_path(modulename='wx.metadata', dirname='mdlib')
+from mdlib.cswlib import CSWBrowserPanel, CSWConnectionPanel
 
 import wx
```
Comment (neteler) [inline grass7/gui/wxpython/wx.metadata/g.gui.cswbrowser/g.gui.cswbrowser.py:20]:
```suggestion
import wx

#%module
#% description: Graphical CSW metadata browser.
#% keyword: general
#% keyword: GUI
#% keyword: metadata
#%end
```

Silence: "WARNING: Bug in UI description. Missing module description"

--> might be extended

**Comment by neteler** [comment]
I started with testing and got some issues:

My setup:
- `python3 dependency.py` runs without error
- Fedora 32
- Python 3.8.3

## Tests with nc_spm_08_grass7 data ([zip](https://grass.osgeo.org/sampledata/north_carolina/nc_spm_08_grass7.zip)):
```
# r.info.iso:
r.info.iso elevation
date of creation: unknown date format
Metadata file has been exported
Traceback (most recent call last):
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/r.info.iso", line 79, in <module>
    main()
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/r.info.iso", line 73, in main
    md.readXML(xml_file)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdgrass.py", line 485, in readXML
    self.md = util.MD_MetadataMOD(etree.parse(xml_file))
AttributeError: module 'owslib.util' has no attribute 'MD_MetadataMOD'


# t.info.iso:
g.extension r.sun.daily
g.region raster=elevation  res=1000 -pa
r.sun.daily elevation=elevation start_day=30 end_day=40 \
            beam_rad_basename=beam beam_rad=beam_sum nprocs=4 -t
# show information about newly created space time dataset
t.info beam

t.info.iso beam
Traceback (most recent call last):
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/t.info", line 128, in <module>
    main()
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/t.info", line 110, in main
    dataset = tgis.dataset_factory(type_, id_)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/factory.py", line 36, in dataset_factory
    sp = SpaceTimeRasterDataset(id)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/space_time_datasets.py", line 1049, in __init__
    AbstractSpaceTimeDataset.__init__(self, ident)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/abstract_space_time_dataset.py", line 52, in __init__
    self.reset(ident)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/space_time_datasets.py", line 1115, in reset
    self.base = STRDSBase(ident=ident)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/base.py", line 934, in __init__
    STDSBase.__init__(self, "strds_base", ident, name, mapset,
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/base.py", line 871, in __init__
    DatasetBase.__init__(self, table, ident, name, mapset, creator,
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/base.py", line 592, in __init__
    self.set_id(ident)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/temporal/base.py", line 617, in set_id
    name, mapset = ident.split("@")
ValueError: too many values to unpack (expected 2)
Traceback (most recent call last):
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/t.info.iso", line 82, in <module>
    sys.exit(main())
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/t.info.iso", line 73, in main
    md = GrassMD(id_, type=type_)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdgrass.py", line 80, in __init__
    self.parseTemporal()
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdgrass.py", line 92, in parseTemporal
    tinfo = Module('t.info',
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 592, in __init__
    self.__call__(*args, **kargs)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 650, in __call__
    return self.run()
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 771, in run
    self.wait()
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/python/grass/pygrass/modules/interface/module.py", line 792, in wait
    raise CalledModuleError(returncode=self.popen.returncode,
grass.exceptions.CalledModuleError: Module run t.info t.info input=beam@user1@user1 type=strds -g ended with error
Process ended with non-zero return code 1. See errors in the (error) output.


# v.info.iso:
v.info.iso schools_wake
Metadata file exists:
/home/mneteler/grassdata/nc_spm_08_grass7/user1/metadata/vector_schools_wake.xml
Metadata file has been overwritten
Traceback (most recent call last):
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/v.info.iso", line 78, in <module>
    sys.exit(main())
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/scripts/v.info.iso", line 72, in main
    md.readXML(xml_file)
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/mdgrass.py", line 485, in readXML
    self.md = util.MD_MetadataMOD(etree.parse(xml_file))
AttributeError: module 'owslib.util' has no attribute 'MD_MetadataMOD'
```

**Comment by neteler** [comment]
More testing:

# GUI
`g.gui.cswbrowser`
- sidenote: many registered URLs are outdated, that's a separate TODO
- to test, select "Finland: National CSW (Paikkatietohakemisto)"
- keyword: "elevation"
- result: 85 results

Wish: change to "Busy" cursor when clicking on the "search" button (I guess somewhere near `wx.metadata/mdlib/cswlib.py`):
```
     wx.BeginBusyCursor()
     ... search ...
     wx.EndBusyCursor()
```


**Bug**: when clicking on one of the datasets listed under "results":

```
(g.gui.cswbrowser:306043): Gtk-CRITICAL **: 11:28:28.938: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkSpinButton
Traceback (most recent call last):
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/cswlib.py", line 376, in OnRecord
    metadata = render_template('en', self.context,
  File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/cswutil.py", line 55, in render_template
    template = env.get_template(template_file)
  File "/usr/lib/python3.8/site-packages/jinja2/environment.py", line 883, in get_template
    return self._load_template(name, self.make_globals(globals))
  File "/usr/lib/python3.8/site-packages/jinja2/environment.py", line 857, in _load_template
    template = self.loader.load(self, name, globals)
  File "/usr/lib/python3.8/site-packages/jinja2/loaders.py", line 115, in load
    source, filename, uptodate = self.get_source(environment, name)
  File "/usr/lib/python3.8/site-packages/jinja2/loaders.py", line 197, in get_source
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: record_metadata_dc.html
```
The missing file is indeed installed in `etc/wx.metadata/config/record_metadata_dc.html` but apparently not found.
Seems the related code is here:

https://github.com/OSGeo/grass-addons/blob/7d2dd42b66beb089a3a635fded7b37ca46419577/grass7/gui/wxpython/wx.metadata/mdlib/cswlib.py#L333

**Comment by neteler** [comment]
More testing:

`g.gui.metadata`

(following https://grasswiki.osgeo.org/wiki/ISO/INSPIRE_Metadata_Support#GRASS_metadata_map_editor and https://www.youtube.com/watch?v=7V5NyVdO05o)

## Initial layout issue

When opening `g.gui.metadata` then the metadata profile cannot be well seen:

image

After a window resize is becomes right:

image

# Usage

Next step is to select a map from the layer tree and click on "Template". This will open a new dialog.

Here I get (not sure what it means to me):

https://github.com/OSGeo/grass-addons/blob/7d2dd42b66beb089a3a635fded7b37ca46419577/grass7/gui/wxpython/wx.metadata/mdlib/mdeditorfactory.py#L888

Next one gets into the editing mode to populate the metadata fields. Here also the layout isn't fully right (resizing the window fixes it by apparently refreshing it):

image

to then become the expected:

image

(I'll continue later with testing)

File: grass7/gui/wxpython/wx.metadata/g.gui.metadata/g.gui.metadata.py (hunk 1 above)
Comment (tmszi) [reply grass7/gui/wxpython/wx.metadata/g.gui.metadata/g.gui.metadata.py:63]: I fixed it.

**Comment by tmszi** [comment]
> Thanks for your enormous efforts to update this suite!
> 
> Just two minor warnings I got, which might be fixed as well:
> 

You are welcome. I fixed it.

File: grass7/gui/wxpython/wx.metadata/g.gui.cswbrowser/g.gui.cswbrowser.py (hunk 2 above)
Comment (tmszi) [reply grass7/gui/wxpython/wx.metadata/g.gui.cswbrowser/g.gui.cswbrowser.py:27]: I fixed it.

**Comment by tmszi** [comment]
> I started with testing and got some issues:
> 
> My setup:
> 
>     * `python3 dependency.py` runs without error
> 
>     * Fedora 32
> 
>     * Python 3.8.3
> 
> 

I fixed it (bugs related with launching `r.info.iso`, `v.info.iso`, `t.info.iso` module).

**Comment by tmszi** [comment]
> More testing:
> # GUI
> 
> `g.gui.cswbrowser`
> 
>     * sidenote: many registered URLs are outdated, that's a separate TODO, see #229
>       
>       * maybe helpful: https://live.osgeo.org/en/quickstart/pycsw_quickstart.html
> 
>     * to test, select "Finland: National CSW (Paikkatietohakemisto)"
> 
>     * keyword: "elevation"
> 
>     * result: 85 results
> 
> 
> Wish: change to "Busy" cursor when clicking on the "search" button (I guess somewhere near `wx.metadata/mdlib/cswlib.py`):
> 
> ```
>      wx.BeginBusyCursor()
>      ... search ...
>      wx.EndBusyCursor()
> ```

I Implemented it (busy cursor).

> 
> **Bug**: when clicking on one of the datasets listed under "results":
> 
> ```
> (g.gui.cswbrowser:306043): Gtk-CRITICAL **: 11:28:28.938: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkSpinButton
> Traceback (most recent call last):
>   File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/cswlib.py", line 376, in OnRecord
>     metadata = render_template('en', self.context,
>   File "/home/mneteler/software/grass78_git/dist.x86_64-pc-linux-gnu/etc/wx.metadata/mdlib/cswutil.py", line 55, in render_template
>     template = env.get_template(template_file)
>   File "/usr/lib/python3.8/site-packages/jinja2/environment.py", line 883, in get_template
>     return self._load_template(name, self.make_globals(globals))
>   File "/usr/lib/python3.8/site-packages/jinja2/environment.py", line 857, in _load_template
>     template = self.loader.load(self, name, globals)
>   File "/usr/lib/python3.8/site-packages/jinja2/loaders.py", line 115, in load
>     source, filename, uptodate = self.get_source(environment, name)
>   File "/usr/lib/python3.8/site-packages/jinja2/loaders.py", line 197, in get_source
>     raise TemplateNotFound(template)
> jinja2.exceptions.TemplateNotFound: record_metadata_dc.html
> ```
> 
> The missing file is indeed installed in `etc/wx.metadata/config/record_metadata_dc.html` but apparently not found.
> Seems the related code is here:
> 
> https://github.com/OSGeo/grass-addons/blob/7d2dd42b66beb089a3a635fded7b37ca46419577/grass7/gui/wxpython/wx.metadata/mdlib/cswlib.py#L333

I can't reproduce it. I got following error only.

g_gui_cswbrowser

**Comment by neteler** [comment]
@tmszi - what about merging for now and then continue in a separate PR? That might help with testing by others being less familiar with locally applying PR changes... (just an idea).

**Comment by tmszi** [comment]
> @tmszi - what about merging for now and then continue in a separate PR? That might help with testing by others being less familiar with locally applying PR changes... (just an idea).

Good idea, I merged it.

**Comment by neteler** [comment]
> Good idea, I merged it.

It is still open and not yet merged ... :-) on purpose?

**Comment by tmszi** [comment]
> > Good idea, I merged it.
> 
> It is still open and not yet merged ... :-) on purpose?

Yes, you are right. I just noticed (as you alerted me) that the only allowed merge option is **squash and merge**. I tried **rebase and merge** via cmd  and than push. It did not print any warning that the option is not allowed. Should I do it through the squash and merge option?

**Comment by neteler** [comment]
Yes, please.

**Pull request merged**